File: Merge Sort/MergeSort-Go/MergeSort.go
package main

import "fmt"

// custom type array
type MyArray []int64

func (arr MyArray) MergeSort() {
	arr.sort(0, len(arr)-1)
}

func (arr MyArray) sort(left, right int) {
	if left < right {
		var mid int = (left + right) / 2
		arr.sort(left, mid)
		arr.sort(mid+1, right)
		arr.merge(left, mid, right)
	}
}

func (arr MyArray) merge(left, mid, right int) {
	n1, n2 := mid-left+1, right-mid
	lefts := make([]int64, n1)
	rights := make([]int64, n2)
	for i := 0; i < n1; i++ {
		lefts[i] = arr[left+i]
	}
	for j := 0; j < n2; j++ {
		rights[j] = arr[mid+1+j]
	}

	i, j, k := 0, 0, left
	for i < n1 && j < n2 {
		if lefts[i] <= rights[j] {
			arr[k] = lefts[i]
			i++
		} else {
			arr[k] = rights[j]
			j++
		}
		k++
	}

	for i < n1 {
		arr[k] = lefts[i]
		i++
		k++
	}

	for j < n2 {
		arr[k] = rights[j]
		j++
		k++
	}
}

func main() {
	arr := MyArray{90, 30, 5, 3, 15, 1, 2}
	fmt.Printf("Before sort: %v\n", arr) //Before sort: [90 30 5 3 15 1 2]
	arr.MergeSort()
	fmt.Printf("After sort: %v\n", arr) //After sort: [1 2 3 5 15 30 90]
}

File: Merge Sort/MergeSort-Dart/MergeSort.dart
void main() {
  List<int> a = [1, 2, 88, 54, -5, 90, 60, 3, 4];
  print(mergeSort(a));
}


List<int> mergeSort(List<int> numbers) {
  int length = numbers.length;
  if(length < 2) {
    return numbers;
  }
  int mid = (length / 2).floor();
  List<int> left = new List(mid);
  List<int> right = new List(length - mid);
  
  for (int i = 0; i < mid; i++) {
    left[i] = numbers[i];
  }
  for (int i = mid; i < length; i++) {
    right[i - mid] = numbers[i];
  }
  left = mergeSort(left);
  right = mergeSort(right);
  
  return merge(left, right, numbers);
}

List<int> merge(List<int> left, List<int> right, List<int> A) {
  int lengthLeft = left.length;
  int lengthRight = right.length;
  int i = 0;
  int j = 0;
  int k = 0;
  
  while (i < lengthLeft && j < lengthRight) {
    if (left[i] <= right[j]) {
      A[k] = left[i];
      i += 1;
    }  else {
      A[k] = right[j];
      j += 1;
    }
    k += 1;
  }
  
  while (i < lengthLeft) {
    A[k] = left[i];
    i += 1;
    k += 1;
  }
  
  while (j < lengthRight) {
    A[k] = right[j];
    j += 1;
    k += 1;
  }
  return A;
}